File: src/com/crdroid/settings/crDroidSettingsLayout.java
/*
 * Copyright (C) 2017-2022 crDroid Android Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.crdroid.settings;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.LinearLayout;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.android.internal.logging.nano.MetricsProto;
import com.android.settings.R;
import com.android.settings.SettingsPreferenceFragment;

import com.crdroid.settings.fragments.About;
import com.crdroid.settings.fragments.Buttons;
import com.crdroid.settings.fragments.LockScreen;
import com.crdroid.settings.fragments.Miscellaneous;
import com.crdroid.settings.fragments.Navigation;
import com.crdroid.settings.fragments.Notifications;
import com.crdroid.settings.fragments.QuickSettings;
import com.crdroid.settings.fragments.Recents;
import com.crdroid.settings.fragments.Sound;
import com.crdroid.settings.fragments.StatusBar;
import com.crdroid.settings.fragments.UserInterface;

public class crDroidSettingsLayout extends SettingsPreferenceFragment {

    private static final String TAG = "crDroidSettingsLayout";
    ViewPager mViewPager;
    ViewGroup mContainer;
    PagerSlidingTabStrip mTabs;
    SectionsPagerAdapter mSectionsPagerAdapter;
    protected Context mContext;

    private static final int MENU_RESET = Menu.FIRST;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getActivity().setTitle(R.string.crdroid_settings_title);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        mContainer = container;
        View view = inflater.inflate(R.layout.crdroid_settings, container, false);
        mViewPager = (ViewPager) view.findViewById(R.id.viewpager);
        mTabs = (PagerSlidingTabStrip) view.findViewById(R.id.tabs);
        mSectionsPagerAdapter = new SectionsPagerAdapter(getFragmentManager());
        mViewPager.setAdapter(mSectionsPagerAdapter);
        mViewPager.setClipChildren(true);
        mViewPager.setClipToPadding(true);
        mTabs.setViewPager(mViewPager);
        mContext = getActivity().getApplicationContext();

        return view;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
    }

    @Override
    public void onSaveInstanceState(Bundle saveState) {
        super.onSaveInstanceState(saveState);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        menu.add(0, MENU_RESET, 0, R.string.reset_settings_title)
                .setIcon(R.drawable.ic_reset)
                .setAlphabeticShortcut('r')
                .setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM |
                        MenuItem.SHOW_AS_ACTION_WITH_TEXT);
    }

    public void resetAll(Context context) {
        new ResetAllTask(context).execute();
    }

    public void showResetAllDialog(Context context) {
        new AlertDialog.Builder(context)
                .setTitle(R.string.reset_settings_title)
                .setMessage(R.string.reset_settings_message)
                .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        resetAll(context);
                    }
                })
                .setNegativeButton(R.string.cancel, null)
                .show();
    }

    private class ResetAllTask extends AsyncTask<Void, Void, Void> {
        private Context rContext;

        public ResetAllTask(Context context) {
            super();
            rContext = context;
        }

        @Override
        protected Void doInBackground(Void... params) {
            Buttons.reset(rContext);
            LockScreen.reset(rContext);
            Miscellaneous.reset(rContext);
            Navigation.reset(rContext);
            Notifications.reset(rContext);
            QuickSettings.reset(rContext);
            Recents.reset(rContext);
            Sound.reset(rContext);
            StatusBar.reset(rContext);
            UserInterface.reset(rContext);
            finish();
            startActivity(getIntent());
            return null;
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case MENU_RESET:
                 showResetAllDialog(getActivity());
                return true;
            default:
                return false;
        }
    }

    class SectionsPagerAdapter extends FragmentPagerAdapter {

        String titles[] = getTitles();
        private Fragment frags[] = new Fragment[titles.length];

        public SectionsPagerAdapter(FragmentManager fm) {
            super(fm);
            frags[0] = new StatusBar();
            frags[1] = new QuickSettings();
            frags[2] = new LockScreen();
            frags[3] = new Recents();
            frags[4] = new Navigation();
            frags[5] = new Buttons();
            frags[6] = new UserInterface();
            frags[7] = new Notifications();
            frags[8] = new Sound();
            frags[9] = new Miscellaneous();
            frags[10] = new About();
        }

        @Override
        public Fragment getItem(int position) {
            return frags[position];
        }

        @Override
        public int getCount() {
            return frags.length;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return titles[position];
        }
    }

    private String[] getTitles() {
        String titleString[];
        titleString = new String[] {
            getString(R.string.statusbar_title),
            getString(R.string.quicksettings_title),
            getString(R.string.lockscreen_title),
            getString(R.string.recents_title),
            getString(R.string.navigation_title),
            getString(R.string.button_title),
            getString(R.string.ui_title),
            getString(R.string.notifications_title),
            getString(R.string.sound_title),
            getString(R.string.misc_title),
            getString(R.string.about_crdroid)
        };
        return titleString;
    }

    @Override
    public int getMetricsCategory() {
        return MetricsProto.MetricsEvent.CRDROID_SETTINGS;
     }
}

File: src/com/crdroid/settings/crDroidSettingsLayout.kt
/*
 * SPDX-FileCopyrightText: crDroid Android Project
 * SPDX-License-Identifier: GPL-3.0
 */

package com.crdroid.settings

import android.app.AlertDialog
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup

import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2

import com.android.internal.logging.nano.MetricsProto
import com.android.settings.R
import com.android.settings.SettingsPreferenceFragment

import com.crdroid.settings.fragments.*

import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class crDroidSettingsLayout : SettingsPreferenceFragment() {

    private var pager: ViewPager2? = null
    private var tabs: TabLayout? = null
    private var mediator: TabLayoutMediator? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        requireActivity().setTitle(R.string.crdroid_settings_title)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val root = inflater.inflate(R.layout.crdroid_settings, container, false)

        tabs = root.findViewById(R.id.tabs)
        pager = root.findViewById(R.id.pager)

        val adapter = object : FragmentStateAdapter(this) {
            override fun getItemCount(): Int = TAB_TITLES.size

            override fun createFragment(position: Int): Fragment = when (position) {
                0 -> StatusBar()
                1 -> QuickSettings()
                2 -> LockScreen()
                3 -> Buttons()
                4 -> UserInterface()
                5 -> Notifications()
                6 -> Sound()
                7 -> Miscellaneous()
                8 -> About()
                else -> StatusBar()
            }
        }

        pager!!.adapter = adapter
        pager!!.offscreenPageLimit = TAB_TITLES.size

        mediator = TabLayoutMediator(tabs!!, pager!!) { tab, position ->
            tab.setText(TAB_TITLES[position])
        }.also { it.attach() }

        return root
    }

    override fun onDestroyView() {
        mediator?.detach()
        mediator = null
        pager = null
        tabs = null
        super.onDestroyView()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        menu.add(0, MENU_RESET, 0, R.string.reset_settings_title)
            .setIcon(R.drawable.ic_reset)
            .setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == MENU_RESET) {
            showResetAllDialog(requireContext())
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun showResetAllDialog(context: Context) {
        AlertDialog.Builder(context)
            .setTitle(R.string.reset_settings_title)
            .setMessage(R.string.reset_settings_message)
            .setPositiveButton(R.string.ok) { _, _ -> resetAll(context) }
            .setNegativeButton(R.string.cancel, null)
            .show()
    }

    private fun resetAll(context: Context) {
        lifecycleScope.launch(Dispatchers.Default) {
            Buttons.reset(context)
            LockScreen.reset(context)
            Miscellaneous.reset(context)
            Navigation.reset(context)
            Notifications.reset(context)
            QuickSettings.reset(context)
            Sound.reset(context)
            StatusBar.reset(context)
            UserInterface.reset(context)

            withContext(Dispatchers.Main) {
                requireActivity().finish()
                startActivity(requireActivity().intent)
            }
        }
    }

    override fun getMetricsCategory(): Int =
        MetricsProto.MetricsEvent.CRDROID_SETTINGS

    companion object {
        private const val MENU_RESET = Menu.FIRST

        private val TAB_TITLES = intArrayOf(
            R.string.statusbar_title,
            R.string.quicksettings_title,
            R.string.lockscreen_title,
            R.string.button_title,
            R.string.ui_title,
            R.string.notifications_title,
            R.string.sound_title,
            R.string.misc_title,
            R.string.about_crdroid
        )
    }
}
